### ISSUE TITLE
emacsql urges to install sqlite3 for magit what crashes Emacs

### ISSUE THREAD

**Issue by vitaminace33** [pekingduck/emacs-sqlite3-api#9]
I swim upstream from here syl20bnr/spacemacs#15992. Read at the end for extra details.

#### Description :octocat:
I recently performed a clean install of Spacemacs (because of reasons) with `rm -rf ~/.emacs.d && git clone [...] && emacs`. After that, when accessing the status of a git repository, a message would urge me to install `sqlite3` (reported at [r/spacemacs](https://www.reddit.com/r/spacemacs/comments/12w4mc3/sqlite_error_with_emacsql_and_forge/)). Doing so and issuing `M-m g s` crashes Emacs.

#### Reproduction guide :beetle:
1. Clean install Spacemacs with `git` layer enabled
2. Start Emacs (everything is installed)
3. Install `sqlite3`
   - OS' dynamic library: in a shell type `sudo apt install libsqlite3-{0,dev}`
   - Emacs' pacakge, add `sqlite3` to `dotspacemacs-additional-packages` (`M-m f e d`)
4. Restart Emacs, issue `M-m q R`
   - If not installed yet, Emacs will download `sqlite3`
5. Open a git repository
6. Check status, issue `M-m g s`
   - If not built yet, Emacs will compile `sqlite3.el`
7. Repeat steps 4-5-6

*Observed behaviour:* :eyes: :broken_heart:
- Emacs should crash after completing step 7.
- If in a first round steps 3-4 are omitted, `emacsql` will eventually print the the following message.
```
Warning (emacsql): Since your Emacs does not come with
built-in SQLite support [1], but does support C modules, the best
EmacSQL backend is provided by the third-party `sqlite3' package
[2].

Please install the `sqlite3' Elisp package using your preferred
Emacs package manager, and install the SQLite shared library
using your distribution's package manager.  That package should
be named something like `libsqlite3' [3] and NOT just `sqlite3'.

In the current Emacs instance the legacy backend is used, which
uses a custom SQLite executable.  Using an external process like
that is less reliable and less performant, and in a few releases
support for that might be removed.

[1]: Supported since Emacs 29.1, provided it was not disabled
     with `--without-sqlite3'.
[2]: https://github.com/pekingduck/emacs-sqlite3-api
[3]: On Debian https://packages.debian.org/buster/libsqlite3-0 Disable showing Disable logging
```

*Expected behaviour:* :heart: :smile:
- Emacs should not crash.
- Installation of `sqlite3` should be taken care of by the `git` layer or at least be documented.
  - In this sense,`libsqlite3-dev` is only needed at compile time.

#### System Info :computer:
- sqlite3: libsqlite3-dev 3.37.2-2ubuntu0.1
- OS: Ubuntu 22.04 LTS
- Emacs: 28.1
- Spacemacs: 0.999.0
- Spacemacs branch: develop (rev. 0bbb41ec9)
- Graphic display: t
- Running in daemon: nil
- Distribution: spacemacs
- Editing style: emacs
- Completion: helm
- Layers:
```elisp
(auto-completion better-defaults bibtex
                 (c-c++ :variables c-c++-backend 'lsp-clangd)
                 emacs-lisp epub git helm html
                 (julia :variables julia-backend 'lsp julia-repl-switches "--threads 8")
                 (latex :variables latex-enable-auto-fill nil latex-enable-magic nil)
                 (lsp :variables lsp-julia-default-environment "~/.julia/environments/v1.8")
                 markdown mu4e multiple-cursors octave org pdf
                 (shell :variables shell-default-position 'right shell-default-shell 'vterm)
                 shell-scripts spacemacs-org spell-checking syntax-checking version-control treemacs
                 (yaml :variables yaml-enable-lsp t))
```
- System configuration features: ACL CAIRO DBUS FREETYPE GIF GLIB GMP GNUTLS GPM GSETTINGS HARFBUZZ JPEG JSON LCMS2 LIBOTF LIBSELINUX LIBSYSTEMD LIBXML2 M17N_FLT MODULES NOTIFY INOTIFY PDUMPER PNG RSVG SECCOMP SOUND THREADS TIFF TOOLKIT_SCROLL_BARS X11 XDBE XIM XPM XWIDGETS GTK3 ZLIB

#### Backtrace :paw_prints:
I don't know.

#### Extra details :heavy_plus_sign:
Under Arch Linux / Emacs 28.2, things are ok. Therefore, I guess there must be something distro related. Moreover, pay attention to the steps to reproduce: at the first run, there is no crash. I have to go through step 7 to trigger the crash, which means to have to restart Emacs after sqlite3-api.so has been built.

**Comment by bryce-carson** [comment]
Crash is reproducible on Fedora 38. When I ignored the warning everything was fine. Installing `sqlite-devel` through DNF and then including the package mentioned in the warning leads to the same crashing behaviour described above.

I am also using the latest development commit of Spacemacs.

**Comment by pekingduck** [comment]
@vitaminace33 I am not a Spacemacs nor a Linux user but I just followed 1-7 on my Mac but wasn't able to reproduce the crash. I guess I'll need to find a Linux machine to further investigate..

**Comment by vitaminace33** [comment]
@pekingduck Try with an Ubuntu (or Fedora) distro.

**Comment by dannywillems** [comment]
Same thing here. On Spacemacs, Ubuntu 22.04

**Comment by brc** [comment]
Really hard for the average user to discern whether this is a Spacemacs problem, Magit/Forge/EmacSQL problem, or emacs-sqlite3-api problem. :)

**TL;DR:** No solution here, just observations about my symptoms.

Spacemacs with Emacs 28.2 on Arch Linux and `sqlite3-20230315.1221` doesn't crash (as already stated by OP), but invoking `(magit-status)` for the first time in any new Emacs session will cause a delay and high CPU usage on my system, followed by this message in the echo area:
```
Database Error: "sqlite_open_v2() failed", 14
```
Nothing happens after that (Magit status buffer is not shown), but invoking `(magit-status)` again immediately shows the status buffer.

Error `14` is [SQLITE_CANTOPEN](https://www.sqlite.org/rescode.html#cantopen), which means:
> SQLite was unable to open a file. The file in question might be a primary database file or one of several [temporary disk files](https://www.sqlite.org/tempfiles.html).
 
This of course could be caused any number of local conditions, including file permissions. I don't know exactly which file Magit is trying to open, but if I introspect the source code on my system, I see that `/path/to/spacemacs/layers/+source-control/git/packages.el` is setting the variable `forge-database-file` to `~/.emacs.d/.cache/forge-database.sqlite`, which exists in my filesystem and has sane ownership/permissions; I can query this database with the `sqlite3(1)` CLI.

It appears to me that Magit doesn't actually use emacsql (or any kind of SQL)&#8212;only Forge does. However, reading other GitHub issues for Magit and grepping the source code reveals that Magit will make use of Forge when present.

I'm not directly or purposefully a user of Forge, though it's installed on my system because I have the `git` layer enabled in Spacemacs.

**Comment by vitaminace33** [comment]
Hi!

I tried again with `libsqlite3-dev:amd64 (3.37.2-2ubuntu0.1)` and `sqlite3-20230315.1221`, that is, the exact same versions I was using earlier, and now it is not crashing. I guess some other update has solved the problem. I'll come back after running further tests.

Regards

**Comment by vitaminace33** [comment]
Hi there!

As @invsblduck said, regular users like me are not able to find easily the culprit. It didn't occur to me that magit was to blame.

Anyhow, I've been using magit with sqlite3 for more than a month with no problem whatsoever. I consider the issue closed.

Thanks for the comments and help.
